flaml/integrations/oai/completion.py

```diff
@@ -0,0 +1,441 @@
+            data_early_stop = False  # whether data early stop happens for this n
+            while True:  # data_limit <= data_length
+                # limit the number of data points to avoid rate limit
+                for i in range(prev_data_limit, data_limit):
+                    x = data[i]
+                    params["prompt"] = prompt.format(**x)
+                    response = cls._get_response(params, eval_only)
+                    if response == -1:  # rate limit error, treat as invalid
+                        if prune:
+                            # update unsafe n and prune this config
+                            cls._min_unsafe_n_per_max_tokens[
+                                region_key
+                            ] = unsafe_n = cls._min_unsafe_n_per_max_tokens.get(
+                                region_key, {}
+                            )
+                            unsafe_n[max_tokens] = min(
+                                n, unsafe_n.get(max_tokens, np.inf)
+                            )
+                        result[metric] = 0
+                        result["cost"] = cost
+                        return result
+                    # evaluate the quality of the responses
+                    responses = [r["text"].rstrip() for r in response["choices"]]
+                    n_tokens = (
+                        response["usage"]["completion_tokens"]
+                        if previous_n
+                        else response["usage"]["total_tokens"]
+                    )
+                    query_cost = (
+                        response["usage"]["total_tokens"]
+                        * cls.price1K[config["model"]]
+                        / 1000
+                    )
+                    cls._total_cost += query_cost
+                    cost += query_cost
+                    if (
+                        cls.optimization_budget
+                        and cls._total_cost >= cls.optimization_budget
+                        and not eval_only
+                    ):
+                        # limit the total tuning cost
+                        return {
+                            metric: 0,
+                            "total_cost": cls._total_cost,
+                            "cost": cost,
+                        }
+                    if previous_n:
+                        n_tokens_list[i] += n_tokens
+                        responses_list[i].extend(responses)
+                    else:
+                        n_tokens_list.append(n_tokens)
+                        responses_list.append(responses)
+                n_tokens = np.mean(n_tokens_list[:data_limit])
+                rho = (
+                    (1 - data_limit / data_length) * (1 + 1 / data_limit)
+                    if data_limit << 1 > data_length
+                    else (1 - (data_limit - 1) / data_length)
+                )
+                # Hoeffding-Serfling bound
+                ratio = 0.1 * np.sqrt(rho / data_limit)
```

Comment:
rho and ratio are not consistent with algorithm 1 in the paper. can you check the correctness?

Reply:
Checked. The code is correct.